pkg/bisect/generic/array_test.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package generic
+
+import (
+	"math"
+	"math/rand"
+	"testing"
+
+	"github.com/google/syzkaller/pkg/testutil"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBisectArrayToZero(t *testing.T) {
+	t.Parallel()
+	array := make([]int, 100)
+	ret, err := Array(ArrayConfig[int]{
+		Pred: func(arr []int) (bool, error) {
+			// No elements are needed.
+			return true, nil
+		},
+		Logf: t.Logf,
+	}, array[:50], array[50:])
+	assert.NoError(t, err)
+	assert.Len(t, ret, 0)
+}
+
+func TestBisectArrayFull(t *testing.T) {
+	t.Parallel()
+	array := make([]int, 100)
+	ret, err := Array(ArrayConfig[int]{
+		Pred: func(arr []int) (bool, error) {
+			// All elements are needed.
+			return false, nil
+		},
+		Logf: t.Logf,
+	}, array)
+	assert.NoError(t, err)
+	assert.Equal(t, ret, array)
+}
+
+func TestBisectRandomArray(t *testing.T) {
+	t.Parallel()
+	r := rand.New(testutil.RandSource(t))
```

Comment:
We do that in quite a number of places throughout syzkaller tests. If it fails anywhere (either locally on or ci), that's already a reason to take a closer look.

Do you suggest replacing this with a set of predefined manual tests here?

Reply:
You can specify seed manually. It should be enough.

Reply:
That would already go against the common practice in our project :)

Reply:
IMHO we should do all we can to not have flaky tests. Having a flaky tests just trains people to just ignore test failures and run the tests again. Also you really have to dig into testutil.RandSource to even see, that we print the seed and how to manually force the same seed again. I think it would be better give testutil.RandSource a fixed seed by default and make the random seed opt in.

Reply:
Yes, we do this in a number of places, and, yes, you need to learn this one (that this happens, the seed it printed, you can control it).

The idea was to get the best of both worlds. CI is not flaky, and we get larger coverage. The search space is really huge for most of these cases. A fixed seed will cover only a tiny part of it.

This worked well in practice so far. No significant flakes on CI, and we do detect the long tail of latent bugs in other runs. FWIW I did not train to ignore these flakes :) I looked at each and every of these. These tests generally don't have timing/environment-dependent flakes.

And note that latent bugs can pop up even with fixed seeds. Consider, one changes program mutation logic to read one more random number, this can trigger a latent bug in another part of mutation.

If we make random seeds opt-in, I afraid nobody will ever enable it. So that will require some additional system to run the random mode continuously (fuzzing effectively). This would probably be a better solution. Just need somebody to implement it :)